File: go-agent/internal/schema/schema.go
package schema

import "time"

type Snapshot struct {
	DeviceID             string   `json:"device_id"`
	Hostname             string   `json:"hostname"`
	Platform             string   `json:"platform"`
	Timestamp            string   `json:"timestamp"`
	SourceOK             bool     `json:"source_ok"`
	CPUTemp              *float64 `json:"cpu_temp"`
	CPULoad              *float64 `json:"cpu_load"`
	CPUClock             *float64 `json:"cpu_clock"`
	CPUPower             *float64 `json:"cpu_power"`
	GPUTemp              *float64 `json:"gpu_temp"`
	GPULoad              *float64 `json:"gpu_load"`
	GPUClock             *float64 `json:"gpu_clock"`
	GPUPower             *float64 `json:"gpu_power"`
	MemoryUsedMB         *float64 `json:"memory_used_mb"`
	MemoryTotalMB        *float64 `json:"memory_total_mb"`
	MemoryPercent        *float64 `json:"memory_percent"`
	FanRPMMax            *float64 `json:"fan_rpm_max"`
	FanRPMAvg            *float64 `json:"fan_rpm_avg"`
	DiskTempMax          *float64 `json:"disk_temp_max"`
	DiskActivityPercent  *float64 `json:"disk_activity_percent"`
	NetUpBPS             *float64 `json:"net_up_bps"`
	NetDownBPS           *float64 `json:"net_down_bps"`
	SystemPowerEstimated *float64 `json:"system_power_estimated"`
}

func Empty(deviceID, hostname, platform string) Snapshot {
	return Snapshot{
		DeviceID:  deviceID,
		Hostname:  hostname,
		Platform:  platform,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		SourceOK:  false,
	}
}

File: src/retro_monitor_agent/schema.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from typing import Any, Optional

TELEMETRY_FIELDS = (
    "device_id",
    "hostname",
    "platform",
    "timestamp",
    "source_ok",
    "cpu_temp",
    "cpu_load",
    "cpu_clock",
    "cpu_power",
    "gpu_temp",
    "gpu_load",
    "gpu_clock",
    "gpu_power",
    "memory_used_mb",
    "memory_total_mb",
    "memory_percent",
    "fan_rpm_max",
    "fan_rpm_avg",
    "disk_temp_max",
    "disk_activity_percent",
    "net_up_bps",
    "net_down_bps",
    "system_power_estimated",
)


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


@dataclass
class TelemetrySnapshot:
    device_id: str
    hostname: str
    platform: str
    timestamp: str
    source_ok: bool
    cpu_temp: Optional[float] = None
    cpu_load: Optional[float] = None
    cpu_clock: Optional[float] = None
    cpu_power: Optional[float] = None
    gpu_temp: Optional[float] = None
    gpu_load: Optional[float] = None
    gpu_clock: Optional[float] = None
    gpu_power: Optional[float] = None
    memory_used_mb: Optional[float] = None
    memory_total_mb: Optional[float] = None
    memory_percent: Optional[float] = None
    fan_rpm_max: Optional[float] = None
    fan_rpm_avg: Optional[float] = None
    disk_temp_max: Optional[float] = None
    disk_activity_percent: Optional[float] = None
    net_up_bps: Optional[float] = None
    net_down_bps: Optional[float] = None
    system_power_estimated: Optional[float] = None

    def to_dict(self) -> dict[str, Any]:
        payload = asdict(self)
        return {field: payload[field] for field in TELEMETRY_FIELDS}

    @classmethod
    def empty(cls, *, device_id: str, hostname: str, platform: str, source_ok: bool = False) -> "TelemetrySnapshot":
        return cls(
            device_id=device_id,
            hostname=hostname,
            platform=platform,
            timestamp=utc_now_iso(),
            source_ok=source_ok,
        )
